src/fu-device-list.c

```diff
@@ -0,0 +1,538 @@
+	FuDevice		*device_old;
+	FuDeviceList		*self;		/* no ref */
+	guint			 remove_id;
+} FuDeviceItem;
+
+G_DEFINE_TYPE (FuDeviceList, fu_device_list, G_TYPE_OBJECT)
+
+static void
+fu_device_list_emit_device_added (FuDeviceList *self, FuDevice *device)
+{
+	g_debug ("::added %s", fu_device_get_id (device));
+	g_signal_emit (self, signals[SIGNAL_ADDED], 0, device);
+}
+
+static void
+fu_device_list_emit_device_removed (FuDeviceList *self, FuDevice *device)
+{
+	g_debug ("::removed %s", fu_device_get_id (device));
+	g_signal_emit (self, signals[SIGNAL_REMOVED], 0, device);
+}
+
+static void
+fu_device_list_emit_device_changed (FuDeviceList *self, FuDevice *device)
+{
+	g_debug ("::changed %s", fu_device_get_id (device));
+	g_signal_emit (self, signals[SIGNAL_CHANGED], 0, device);
+}
+
+/**
+ * fu_device_list_get_all:
+ * @self: A #FuDeviceList
+ *
+ * Returns all the devices that have been added to the device list.
+ * This includes devices that are no longer active, for instance where a
+ * different plugin has taken over responsibility of the #FuDevice.
+ *
+ * Returns: (transfer container) (element-type FuDevice): the devices
+ *
+ * Since: 1.0.2
+ **/
+GPtrArray *
+fu_device_list_get_all (FuDeviceList *self)
+{
+	GPtrArray *devices;
+	g_return_val_if_fail (FU_IS_DEVICE_LIST (self), NULL);
+	devices = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+	for (guint i = 0; i < self->devices->len; i++) {
+		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
+		g_ptr_array_add (devices, g_object_ref (item->device));
+	}
+	for (guint i = 0; i < self->devices->len; i++) {
+		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
+		if (item->device_old == NULL)
+			continue;
+		g_ptr_array_add (devices, g_object_ref (item->device_old));
+	}
+	return devices;
+}
+
+/**
+ * fu_device_list_get_active:
+ * @self: A #FuDeviceList
+ *
+ * Returns all the active devices that have been added to the device list.
+ * An active device is defined as a device that is currently conected and has
+ * is owned by a plugin.
+ *
+ * Returns: (transfer container) (element-type FuDevice): the devices
+ *
+ * Since: 1.0.2
+ **/
+GPtrArray *
+fu_device_list_get_active (FuDeviceList *self)
+{
+	GPtrArray *devices;
+	g_return_val_if_fail (FU_IS_DEVICE_LIST (self), NULL);
+	devices = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+	for (guint i = 0; i < self->devices->len; i++) {
+		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
+		g_ptr_array_add (devices, g_object_ref (item->device));
+	}
+	return devices;
+}
+
+static FuDeviceItem *
+fu_device_list_find_by_device (FuDeviceList *self, FuDevice *device)
+{
+	for (guint i = 0; i < self->devices->len; i++) {
+		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
+		if (item->device == device)
+			return item;
+	}
+	for (guint i = 0; i < self->devices->len; i++) {
+		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
+		if (item->device_old == NULL)
```

Comment:
This check only makes sense if the `device` parameter can be `NULL`, right? Otherwise the normal `item->device_old == device` is sufficient. Maybe check for `device == NULL` at the very top of the function?

Reply:
Agreed, fixed, thanks. The static functions don't have any `g_return_val_if_fail`'s deliberately, and I've verified that none of the callers allow `device=NULL`.